### ISSUE TITLE
No internet connectivity from WSL2/Ubuntu

### ISSUE THREAD

**Issue by j0057** [microsoft/WSL#4926]
* Your Windows build number:  10.0.19041.84

* What you're doing and what's happening:  

I'm getting timeouts when trying to connect to internet from a WSL2 instance. DNS works, and the WSL2 instance can see/ping the host without problem.

    $ ip route
    default via 172.24.64.1 dev eth0
    172.24.64.0/20 dev eth0 proto kernel scope link src 172.24.66.230
    $ cat /etc/resolv.conf
    # This file was automatically generated by WSL. To stop automatic generation of this file, add the following entry to /etc/wsl.conf:
    # [network]
    # generateResolvConf = false
    nameserver 172.24.64.1
    $ dig +noall +answer microsoft.com
    microsoft.com.          0       IN      A       13.77.161.179
    microsoft.com.          0       IN      A       40.76.4.15
    microsoft.com.          0       IN      A       40.112.72.205
    microsoft.com.          0       IN      A       40.113.200.201
    microsoft.com.          0       IN      A       104.215.148.63
    $ curl -4sv -m5 https://microsoft.com/
    *   Trying 13.77.161.179...
    * TCP_NODELAY set
    * After 2498ms connect time, move on!
    * connect to 13.77.161.179 port 443 failed: Connection timed out
    *   Trying 40.76.4.15...
    * TCP_NODELAY set
    * After 1249ms connect time, move on!
    * connect to 40.76.4.15 port 443 failed: Connection timed out
    *   Trying 40.112.72.205...
    * TCP_NODELAY set
    * After 623ms connect time, move on!
    * connect to 40.112.72.205 port 443 failed: Connection timed out
    *   Trying 40.113.200.201...
    * TCP_NODELAY set
    * After 311ms connect time, move on!
    * connect to 40.113.200.201 port 443 failed: Connection timed out
    *   Trying 104.215.148.63...
    * TCP_NODELAY set
    * After 155ms connect time, move on!
    * connect to 104.215.148.63 port 443 failed: Connection timed out
    * Failed to connect to microsoft.com port 443: Connection timed out
    * Closing connection 0

Based on what I know about Hyper-V networking, my Internal switch may or may not route packets from the VM to the internet, but in order for packets to be routed back, the traffic from the WSL2 VM will need to be NAT'ed. 

On the host, in Powershell, `Get-NetNat` returns no results:

    PS C:\WINDOWS\system32> Get-NetIPAddress -InterfaceIndex 59 | Format-Table

    ifIndex IPAddress                                       PrefixLength PrefixOrigin SuffixOrigin AddressState PolicyStore
    ------- ---------                                       ------------ ------------ ------------ ------------ -----------
    59      fe80::xxxx:xxxx:xxxx:xxxx%59                              64 WellKnown    Link         Preferred    ActiveStore
    59      172.24.64.1                                               20 Manual       Manual       Preferred    ActiveStore

    PS C:\WINDOWS\system32> Get-VMSwitch -name WSL | Format-List

    Name                                             : WSL
    Id                                               : 32874a54-04c6-4677-a51f-77c5245345a5
    Notes                                            :
    Extensions                                       : {Microsoft Windows Filtering Platform, Microsoft Azure VFP Switch Extension, Microsoft NDIS Capture}
    BandwidthReservationMode                         : Absolute
    PacketDirectEnabled                              : False
    EmbeddedTeamingEnabled                           : False
    IovEnabled                                       : False
    SwitchType                                       : Internal
    AllowManagementOS                                : True
    NetAdapterInterfaceDescription                   :
    NetAdapterInterfaceDescriptions                  :
    NetAdapterInterfaceGuid                          :
    IovSupport                                       : False
    IovSupportReasons                                :
    AvailableIPSecSA                                 : 0
    NumberIPSecSAAllocated                           : 0
    AvailableVMQueues                                : 0
    NumberVmqAllocated                               : 0
    IovQueuePairCount                                : 0
    IovQueuePairsInUse                               : 0
    IovVirtualFunctionCount                          : 0
    IovVirtualFunctionsInUse                         : 0
    PacketDirectInUse                                : False
    DefaultQueueVrssEnabledRequested                 : True
    DefaultQueueVrssEnabled                          : False
    DefaultQueueVmmqEnabledRequested                 : True
    DefaultQueueVmmqEnabled                          : False
    DefaultQueueVrssMaxQueuePairsRequested           : 16
    DefaultQueueVrssMaxQueuePairs                    : 0
    DefaultQueueVrssMinQueuePairsRequested           : 1
    DefaultQueueVrssMinQueuePairs                    : 0
    DefaultQueueVrssQueueSchedulingModeRequested     : StaticVrss
    DefaultQueueVrssQueueSchedulingMode              : Dynamic
    DefaultQueueVrssExcludePrimaryProcessorRequested : False
    DefaultQueueVrssExcludePrimaryProcessor          : False
    SoftwareRscEnabled                               : False
    BandwidthPercentage                              : 0
    DefaultFlowMinimumBandwidthAbsolute              : 0
    DefaultFlowMinimumBandwidthWeight                : 0
    CimSession                                       : CimSession: .
    ComputerName                                     : NB-XXXXXXX
    IsDeleted                                        : False
    DefaultQueueVmmqQueuePairs                       : 0
    DefaultQueueVmmqQueuePairsRequested              : 16

When trying to set up a new NAT rule, I get a seemingly unrelated error:

    PS C:\WINDOWS\system32> New-NetNat -Name "WSLNetworkNAT" -ExternalIPInterfaceAddressPrefix 172.24.64.0/20
    New-NetNat : The parameter is incorrect.
    At line:1 char:1
    + New-NetNat -Name "WSLNetworkNAT" -ExternalIPInterfaceAddressPrefix 17 ...
    + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        + CategoryInfo          : InvalidArgument: (MSFT_NetNat:root/StandardCimv2/MSFT_NetNat) [New-NetNat], CimException
        + FullyQualifiedErrorId : Windows System Error 87,New-NetNat

**Comment by rm-bergmann** [comment]
I have the same problem, however I get this output from `Get-NetNat`:
```
Name                             : NATNetwork2
ExternalIPInterfaceAddressPrefix :
InternalIPInterfaceAddressPrefix : 192.168.0.0/24
IcmpQueryTimeout                 : 30
TcpEstablishedConnectionTimeout  : 1800
TcpTransientConnectionTimeout    : 120
TcpFilteringBehavior             : AddressDependentFiltering
UdpFilteringBehavior             : AddressDependentFiltering
UdpIdleSessionTimeout            : 120
UdpInboundRefresh                : False
Store                            : Local
Active                           : True

Name                             : NATnetwork
ExternalIPInterfaceAddressPrefix :
InternalIPInterfaceAddressPrefix : 192.168.56.102/24
IcmpQueryTimeout                 : 30
TcpEstablishedConnectionTimeout  : 1800
TcpTransientConnectionTimeout    : 120
TcpFilteringBehavior             : AddressDependentFiltering
UdpFilteringBehavior             : AddressDependentFiltering
UdpIdleSessionTimeout            : 120
UdpInboundRefresh                : False
Store                            : Local
Active                           : True

Name                             : MyNATnetwork
ExternalIPInterfaceAddressPrefix :
InternalIPInterfaceAddressPrefix : 172.21.21.0/24
IcmpQueryTimeout                 : 30
TcpEstablishedConnectionTimeout  : 1800
TcpTransientConnectionTimeout    : 120
TcpFilteringBehavior             : AddressDependentFiltering
UdpFilteringBehavior             : AddressDependentFiltering
UdpIdleSessionTimeout            : 120
UdpInboundRefresh                : False
Store                            : Local
Active                           : True
```
$ ip route
```
default via 172.30.160.1 dev eth0
172.30.160.0/20 dev eth0 proto kernel scope link src 172.30.168.189
```

**Comment by j0057** [comment]
A colleague helpfully pointed out that Symantec Endpoint Protection includes a firewall that blocks the requests from the WSL2 context. (Incidentally, Symantec EP is according to themselves "not supported" for workloads that include running VM's in Hyper-V.)

As for the NAT rules -- with the firewall temporarily disabled I can do a request to the internet, but `Get-NetNAT` still does not return anything, so, apparently, **these NAT rules are not needed for WSL2.** I'm now wondering how it does work under the hood.

**Comment by j0057** [comment]
@rm-bergmann Those NAT rules you're seeing could be related to other virtual switches. You should be able to find out using `Get-NetIPAddress | Format-Table` and `Get-NetAdapter` cmdlets -- you can correlate the `interfaceIndex` fields.

**Comment by rm-bergmann** [comment]
@j0057 Thank you, you have helped me solve my problem that I have been debugging for over 3 months! Those NAT rules that I had there (probably from previous VM's) were conflicting with WSL2, so I removed them with `Remove-NetNat` and boom! WSL2 networking is fixed (in my case)!

**Comment by natronkeltner** [comment]
I ran into this for much of today on 10.0.19546. I have no NetNATs, have Windows AV, and disabled the firewall for testing. WSL1 works fine, WSL2 does not.

From inside WSL2 I could ping my host interfaces and DNS worked fine, but nothing routed.

I experimented with removing options, explicitly setting interface metrics, uninstalling and reinstalling docker, reboots, but nothing worked. I eventually found this partial workaround:

1) Manually enable Internet Connection Sharing: right click the network adapter you want to share, properties, sharing, then enable it for `vEthernet (WSL)`.

![image](https://user-images.githubusercontent.com/4664270/75299268-3ae8d780-57fa-11ea-93be-be3a9df99ae8.png)

2) The IP address for WSL now changed to some default ICS IP, `192.168.137.1`. In WSL2 Ubuntu, edit the eth0 interface to be in the same subnet: ` ip addr add 192.168.137.10/24 dev eth0`

3) Change the default gateway (`ip route delete default` and `ip route add default via 192.168.137.1`)

4) Change DNS (manually edit /etc/resolv.conf)

And it finally had internet. It goes back to defaults on wsl2 shutdown, of course. This also doesn't handle moving from wired to wireless.

Something is definitely broken in the Hyper-V NAT / Routing pieces for vEthernet (WSL), but I couldn't figure it out.

**Comment by j0057** [comment]
@natronkeltner Small data point: when I navigate to the `Sharing` tab of both my uplink adapter as well as the vEthernet adapter, they say _"Internet Connection Sharing has been disabled by the Network Administrator"_. So I'm not even sure how networking is supposed to work in WSL2!

**Comment by Sidneys1** [comment]
So I went through this entire thread as well as #4731 with no solution to my problem, which is slightly different:

I *am* able to **ping** internet resources, such as `google.com` or `1.1.1.1`, but I can't access them with `wget` or `curl` (e.g., `curl -4v google.com`).

I went so far as to uninstall `Ubuntu-18.04`, disable the Windows Feature `Windows Subsystem for Linux`, then re-enable/re-install. I've messed with the network address and routing settings. I've messed with Hyper-V adaptors. I've messed with [Windows reserved ports and dynamic port ranges](https://github.com/docker/for-win/issues/3171#issuecomment-572571882) which have been an issue for me in the past. Still no-go. Some output:

```
$ ping 1.1.1.1
PING 1.1.1.1 (1.1.1.1) 56(84) bytes of data.
64 bytes from 1.1.1.1: icmp_seq=1 ttl=56 time=55.9 ms
$ curl -4v 1.1.1.1
* Rebuilt URL to: 1.1.1.1/
*   Trying 1.1.1.1...
* TCP_NODELAY set
^C
```

So I thought, what else is a lightweight utility VM? Windows Sandbox! So I opened that and in there I can ping, but not browse (with edge) as well... However with `curl.exe` I can get `google.com`'s redirect page to `www.google.com` but not `www.google.com` itself... :(

**Comment by rm-bergmann** [comment]
I eventually found the underlying problem I had which caused my networking to break in WSL2. My previous fix (hack) to set up a network bridge didn't fix my underlying problem. I found that I had some NAT adapters from previous VM's that were set up in Hyper-V. Removing them fixed all my network issues. 

In powershell I ran the cmd `Get-NetNat` and there were 3 NAT's listed. I shutdown WSL and I ran the `Remove-NetNat` cmd, which restored my networking and internet connectivity in WSL2.

**Comment by Sidneys1** [comment]
Yeah, I tried that - I had no `NetNat` objects listed.

I should add that I'm on the latest insider build and have the Docker WSL engine installed, but I disabled that as part of hunting down this issue.

~Sidney Borne

________________________________

**Comment by Sidneys1** [comment]
I finally found the underlying problem! I installed Wireshark and attached to the WSL vSwitch to see if I could diagnose the problem there and...

Turns out all the Ping replies were coming from the same MAC address - and that MAC belonged to an OpenVPN TAP NIC on my host. I uninstalled the NIC and `wsl --shutdown`/restarted and networking works again!

It looks like WSL2 connects to the "best looking" NIC on the host, even if that NIC isn't network-connected. They should add a wsl.conf setting to control what NIC WSL2 gets attached to.

**Comment by luxzg** [comment]
They should add a wsl.conf setting to control what NIC WSL2 gets attached to... and do that for each WSL2 distro, and allow not to use bridge, and... it is still going to worldwide distribution "as is" apparently as W10 2004. I had high hopes for WSL2, but as it stands now, I'm better off keep having few Linux VMs on my local Hyper-V. At least I know which VM is tied to which adapter, which VLAN, which subnet, and which (static!) IP.

**Comment by natronkeltner** [comment]
It took forever to figure this out so I'll document it here in case someone comes along and finds this. I used Microsoft Message Analyzer with the Hyper-V-VmSwitch and NDIS-PacketCapture providers, which will show exactly how the internal switches are routing packets.

![image](https://user-images.githubusercontent.com/4664270/77210932-3e594100-6ad0-11ea-9f3e-8709a61fb520.png)

This allowed me to see the outbound TCP packets, the inbound TCP responses, and that they were being filtered at the switch level due to a VLAN tag.

```
MessageNumber	DiagnosisTypes	Timestamp	TimeElapsed	Source	Destination	Module	Summary	
502	None	2020-03-20T16:21:09.3340883				Microsoft_Windows_Hyper_V_VmSwitch	NBL 0xFFFF90079978BA40 received from Nic 14B75DAC-3764-4D7F-84B6-8F39004014B2 (Friendly Name: WSL) in switch D57EE459-57B6-4965-B86E-1906ABAECB70 (Friendly Name: WSL)	
503	None	2020-03-20T16:21:09.3340929				Microsoft_Windows_Hyper_V_VmSwitch	NBL originating from Nic 14B75DAC-3764-4D7F-84B6-8F39004014B2 (Friendly Name: WSL) was dropped in switch D57EE459-57B6-4965-B86E-1906ABAECB70 (Friendly Name: WSL), Reason VLAN Filtered (Status:Unknown NTSTATUS Error code (0xE0000003))	
```

>`... was dropped in switch D57EE459-57B6-4965-B86E-1906ABAECB70 (Friendly Name: WSL), Reason VLAN Filtered`

My network port was slightly misconfigured and VLAN tagged packets were being sent to my system, but I had never noticed before because nothing connected to this switch port had ever cared before. Windows didn't care (and didn't show me, either, when using Wireshark/npcap). WSL1 also didn't care that inbound packets were tagged. WSL2, however, uses hyper-v switches, which *do* care about VLAN tags and silently drop packets that are tagged.

Fixing my network port to strip packets of VLAN tags fixed my issue and the hyper-v switches now work fine.

**Comment by mitsuka** [comment]
In my environment, I disabled everything in the BIOS except for the interface connected to the internet.
After that, the internet connectivity was restored.

**Comment by walidshaari** [comment]
my issue was wit AVG Internet security enhanced firewall

**Comment by anthonysomerset** [comment]
For those that use Symantec Endpoint Protection and disabling it works

the specific config that seems to get things working again without a complete disable is to allow unmatched IP traffic as per the attached screenshot
![Screenshot 2020-04-15 15 43 51](https://user-images.githubusercontent.com/340584/79344438-225e8900-7f30-11ea-8d02-5c38a99eaab4.png)

**Comment by dennisameling** [comment]
> Turns out all the Ping replies were coming from the same MAC address - and that MAC belonged to an OpenVPN TAP NIC on my host. I uninstalled the NIC and wsl --shutdown/restarted and networking works again!

Had the same issue here, there was a TAP NIC left from some old OpenVPN installation. Working as expected now 🚀

**Comment by oidualc** [comment]
> For those that use Symantec Endpoint Protection and disabling it works
> 
> the specific config that seems to get things working again without a complete disable is to allow unmatched IP traffic as per the attached screenshot
> ![Screenshot 2020-04-15 15 43 51](https://user-images.githubusercontent.com/340584/79344438-225e8900-7f30-11ea-8d02-5c38a99eaab4.png)

This is unfortunate! In my case in a work PC the damn Symantec Endpoint Protection is blocking the communications for WSL2 (not WSL though..), but on the other end I cannot modify that setting because it's blocked by the organization.
Why this is a problem only for WSL2?
And most importantly, is there anything else that can be done without touching the SEP settings?

**Comment by anthonysomerset** [comment]
Your only options is to downgrade to wsl1 and wait for Symantec to release an update

Or ask your org it team for the setting to be changed temporarily 

Anthony Somerset

Please excuse the brevity or any typos, tapped out from my iPhone

**Comment by loscil412** [comment]
Hi, I solved this issue -- being able to ping addresses but not 'communicate' directly with them (in particular, github.com).

I have avast firewall turned on.
Checking the avast fw settings, a few things stood out:

The adapter description was Hyper-V Virtual Ethernet Adapter #3.

1. My network was set as a public network. I changed it to a private network (which it is :)).

2. In settings (a blue link under firewall profile) I set ALLOW For programs with no defined rules (i.e. programs not listed on the Application Rules page) This setting was defaulted to 'Auto-Decide'

I haven't done too much more at this point than verify that I can curl google, and git pull without issue within Ubuntu on WSL2.

I'm on build 19041.330

**Comment by samnf** [comment]
For me, the "resolution" ended up being to restart my PC 🤦 ..
So make sure you try that before anything too crazy

**Comment by Karlheinzniebuhr** [comment]
Having the same issue, installed brand new ubuntu 20.04 with WSL 2 but cannot ping 8.8.8.8 nor any IP out in the internet from ubuntu

**Comment by onomatopellan** [comment]
@Karlheinzniebuhr make sure you didn't hit this bug #5437

**Comment by Karlheinzniebuhr** [comment]
Fixed by applying this https://github.com/microsoft/WSL/issues/5336#issuecomment-668773304 

=============================================================================
FIX WSL2 NETWORKING IN WINDOWS 10
cmd as admin:
wsl --shutdown
netsh winsock reset
netsh int ip reset all
netsh winhttp reset proxy
ipconfig /flushdns

Windows Search > Network Reset

Restart Windows

**Comment by fkguo** [comment]
> Fixed by applying this [#5336 (comment)](https://github.com/microsoft/WSL/issues/5336#issuecomment-668773304)
> 
> =============================================================================
> FIX WSL2 NETWORKING IN WINDOWS 10
> cmd as admin:
> wsl --shutdown
> netsh winsock reset
> netsh int ip reset all
> netsh winhttp reset proxy
> ipconfig /flushdns
> 
> Windows Search > Network Reset
> 
> Restart Windows

Thank you for the tip. I followed these steps, but unfortunately I completely lost my internet connection even for windows... 

ipconfig /all shows media disconnected, and I tried methods that I found on line, but no one worked... Any clue?

**Comment by Karlheinzniebuhr** [comment]
Do you need to set up a static IP with a specific gateway? I suspect your
lan has no DHCP server and you might need to setup your previous
configuration

**Comment by fkguo** [comment]
> Do you need to set up a static IP with a specific gateway? I suspect your lan has no DHCP server and you might need to setup your previous configuration

Thanks for the quick reply. No, the previous setting was automatic. I don't need to set anything in principle...

One of the steps I followed was to `netsh int ip reset reset.log` as an administrator. One of the resettings failed. Don't know whether this is related or not.

**Comment by fkguo** [comment]
![20200831_215111](https://user-images.githubusercontent.com/62633089/91727341-271c9480-ebd4-11ea-909e-bacc748a15a7.jpg)
One of the output lines says "Resetting , failed. Access is denied."

**Comment by viralrsejpal** [comment]
> 
> 
> Fixed by applying this [#5336 (comment)](https://github.com/microsoft/WSL/issues/5336#issuecomment-668773304)
> 
> =============================================================================
> FIX WSL2 NETWORKING IN WINDOWS 10
> cmd as admin:
> wsl --shutdown
> netsh winsock reset
> netsh int ip reset all
> netsh winhttp reset proxy
> ipconfig /flushdns
> 
> Windows Search > Network Reset
> 
> Restart Windows

I was using Ubuntu on windows, Following the above mentioned steps I also had to Reset the app. 
Steps to Follow:

Windows -> Ubuntu -> App Settings ( context menu/right click ) -> Reset App. 

This will delete your data, I was just setting up this so I did not have any data.
Thank you

**Comment by fandykun** [comment]
Having the same issues after updating windows insiders into `build 20215`. I solved this by restarting my PC.

**Comment by SangNguyenTL** [comment]
> Having the same issues after updating windows insiders into `build 20215`. I solved this by restarting my PC.

Me too. But sometimes it still doesn't work. It is pathetic when playing gambling. 👎

**Comment by raidenii** [comment]
Another way of allowing traffic for people use Symantec Endpoint Protection without allowing all IP traffic:

Add a firewall rule that allows IP traffic (or traffic you want, like IP protocol 1 (ICMP), 6 (TCP) and 17 (UDP)) on the Hyper-V adapter, both directions. You can find which adapter WSL2 uses in Network & Sharing Center (it will say vEthernet (WSL) in its name).

**Comment by daneb** [comment]
I have the same problem and I've lost close to 3 hours of my time today alone. Machine goes to sleep. WSL2 loses internet.
Reset Hyper-V. Shutdown WSL. Reset Winsock. Reset Network. Try setup routing manually. Try use two nics. URRRGH
There are so many of these reported problems and they just seem be getting closed without resolution.

😪😪😥😥😥😥😥😥😥😥😥😥😥😥😥😥😥😥😥😥😥😥

**Comment by matthiasradde** [comment]
For me ping to IP-addresses (even on the internet) are working. But DNS was not working :-(
So I changed my `/etc/resolv.conf` to use 8.8.8.8 as DNS-server and created the `/etc/wsl.conf` as described in `/etc/resolv.conf`
Might be a workaround but now the internet-connection from WSL2 is working again - at least for me. :-)

BTW: I'm using Win10 Build 19041.508 and Ubuntu with WSL2 and no additional Kaspersky/Symantec/whatever and have restarted my system.

**Comment by camilo-agudelo** [comment]
I was having the same problem with WSL2/Ubuntu. Ping did work but general connectivity (e.g. curl or telnet) did not. When I downgraded WSL version to 1 everything worked fine. 

I solved the issue by disabling Hyper-V in Windows Features.

**Comment by deduble** [comment]
In my case just disabling the TAP OpenVPN Adapter worked. thanks to sydneys1 answer.

**Comment by dennisameling** [comment]
For me remvoing the TAP OpenVPN adapter indeed seems to have solved the issue. Fingers crossed it will keep working now 🤞

**Comment by danton721** [comment]
> Windows Search > Network Reset

Worked as a charm. But as my Windows is set in another language, just had a trouble to find this specific Network reset step. To help others: Windows Configurations -> Network -> On the "Status" tab, which should be already opened -> Last button on the page, there is the reset network

**Comment by HiFiPhile** [comment]
If you have multiple NICs try to set metric manually.
I've one GbE NIC for Internet and one 10GbE NIC for internal network, I've tried all workarounds but nothing works.
Finally I saw the metric of 10GbE is lower than GbE, so I set the metric of GbE to 1 and problem solved.

**Comment by ubermanl** [comment]
Facing the same issue here, it's very strange
ip a | grep eth0
```shell
4: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    inet 172.26.231.97/20 brd 172.26.239.255 scope global eth0
```
ip route
```shell
default via 172.26.224.1 dev eth0
172.26.224.0/20 dev eth0 proto kernel scope link src 172.26.231.97
```
ipconfig
```shell
Ethernet adapter vEthernet (Default Switch):

   Connection-specific DNS Suffix  . :
   Link-local IPv6 Address . . . . . : fe80::4410:dab:396c:135d%68
   IPv4 Address. . . . . . . . . . . : 172.31.160.1
   Subnet Mask . . . . . . . . . . . : 255.255.240.0
   Default Gateway . . . . . . . . . :

Ethernet adapter vEthernet (WSL):

   Connection-specific DNS Suffix  . :
   Link-local IPv6 Address . . . . . : fe80::d899:fefa:fc81:2e07%95
   IPv4 Address. . . . . . . . . . . : 172.26.224.1
   Subnet Mask . . . . . . . . . . . : 255.255.240.0
   Default Gateway . . . . . . . . . :
```

winver
![image](https://user-images.githubusercontent.com/1846840/96286545-eb3c6380-0fb6-11eb-9f34-84933bd14509.png)

Debian 10.5 installed thru Ms store.

Can ping the debian network ip from powershell
```shell
Pinging 172.26.231.97 with 32 bytes of data:
Reply from 172.26.231.97: bytes=32 time<1ms TTL=64
Reply from 172.26.231.97: bytes=32 time=1ms TTL=64
Reply from 172.26.231.97: bytes=32 time=1ms TTL=64
```

but the other way, just doesnt work at all
```shell
PING 172.26.224.1 (172.26.224.1) 56(84) bytes of data.
^C
--- 172.26.224.1 ping statistics ---
6 packets transmitted, 0 received, 100% packet loss, time 191ms
```
Tried uninstalling and reinstalling several times, with no success, also setting static routes doesn't work, and tried changing Hyper-V Virtual Switch, but whenever i select my wifi adapter, wifi connection goes down, and a reboot is needed to get it back working.

**Comment by guigoalmeida** [comment]
> Another way of allowing traffic for people use Symantec Endpoint Protection without allowing all IP traffic:
> 
> Add a firewall rule that allows IP traffic (or traffic you want, like IP protocol 1 (ICMP), 6 (TCP) and 17 (UDP)) on the Hyper-V adapter, both directions. You can find which adapter WSL2 uses in Network & Sharing Center (it will say vEthernet (WSL) in its name).

How can one do that? Apparently, symantec endpoint protection firewall rules ('security exception risks') should relate to either 1)  known risks, 2) files, 3) folders, 4) extensions, or 5) web domains. (these might not be the exact terms: I am roughly translating back from my interface, which is in Portuguese). 

i have tried to make an exception to wsl.exe, under Windows/System32, but it apparently did not work properly. How could one make exceptions only for certain types of IP traffic? 

In my case, internet seems to be working rather properly (I can ping google.com with good results), but ubuntu's servers (for apt get update, for instance) do not work, and http addresses within command line instructions do not work either.

**Comment by sboonyakiatACR** [comment]
For people with SEP. I had to add new firewall rule in SEP console to allow both local and remote port 80,443.  Only remote doesn't work, for some reason.

**Comment by ubermanl** [comment]
> For people with SEP. I had to add new firewall rule in SEP console to allow both local and remote port 80,443.  Only remote doesn't work, for some reason. 

I have tried with SEP disabled, but still not working.

**Comment by jsteenri** [comment]
Since I found myself here and then eventually found a workaround elsewhere, sharing that resource as well: https://github.com/microsoft/WSL/issues/4150#issuecomment-504209723

**Comment by hatranpro** [comment]
> A colleague helpfully pointed out that Symantec Endpoint Protection includes a firewall that blocks the requests from the WSL2 context. (Incidentally, Symantec EP is according to themselves "not supported" for workloads that include running VM's in Hyper-V.)
> 
> As for the NAT rules -- with the firewall temporarily disabled I can do a request to the internet, but `Get-NetNAT` still does not return anything, so, apparently, **these NAT rules are not needed for WSL2.** I'm now wondering how it does work under the hood.

I confirm the issue with WSL2 is if Symantec Endpoint Protection running. I was able to connect to the internet after stopping Symantec Endpoint Protection. Thank you.

**Comment by kevingosse** [comment]
I ran into the same problem (ping working, not curl). In my case it seems it was another occurrence of https://github.com/microsoft/WSL/issues/4926#issuecomment-597146860. I couldn't find how to capture packets from the WSL network interface in Wireshark, so I just shutdown WSL, disabled a bunch of network adapters, and ran WSL again, and it worked.

**Comment by rjudin** [comment]
Our team reports about verified workaround https://github.com/sakai135/wsl-vpnkit with window 10 build `1909`, that you could try

**Comment by Adriano-Santtos** [comment]
> Hi, I solved this issue -- being able to ping addresses but not 'communicate' directly with them (in particular, github.com).
> 
> I have avast firewall turned on.
> Checking the avast fw settings, a few things stood out:
> 
> The adapter description was Hyper-V Virtual Ethernet Adapter #3.
> 
> 1. My network was set as a public network. I changed it to a private network (which it is :)).
> 2. In settings (a blue link under firewall profile) I set ALLOW For programs with no defined rules (i.e. programs not listed on the Application Rules page) This setting was defaulted to 'Auto-Decide'
> 
> I haven't done too much more at this point than verify that I can curl google, and git pull without issue within Ubuntu on WSL2.
> 
> I'm on build 19041.330

It worked for me.
Thank you!

**Comment by Adriano-Santtos** [comment]
> Hi, I solved this issue -- being able to ping addresses but not 'communicate' directly with them (in particular, github.com).
> 
> I have avast firewall turned on.
> Checking the avast fw settings, a few things stood out:
> 
> The adapter description was Hyper-V Virtual Ethernet Adapter #3.
> 
> 1. My network was set as a public network. I changed it to a private network (which it is :)).
> 2. In settings (a blue link under firewall profile) I set ALLOW For programs with no defined rules (i.e. programs not listed on the Application Rules page) This setting was defaulted to 'Auto-Decide'
> 
> I haven't done too much more at this point than verify that I can curl google, and git pull without issue within Ubuntu on WSL2.
> 
> I'm on build 19041.330

It worked for me.
Thank you!

**Comment by kdxcxs** [comment]
After trying every single way mentioned above and failed, I finally manage to find a way that works for me. Just run `netsh int ipv4 set dynamic tcp start=49152 num=16384` as Administrator and wsl could reach the internet.

I've set some `Protocol tcp Port Exclusion Ranges` before as Docker and JetBrains' IDEs could not run properly, which might just cause this issue.

**Comment by ambethell** [comment]
> Our team reports about verified workaround https://github.com/sakai135/wsl-vpnkit with window 10 build `1909`, that you could try

I can confirm this helped me get around SEP (14.2.RU2 MP1 build 5569 - 14.2.5569.2100) issues.  The configuration for my version of SEP also had 'allow IP traffic' on unmatched IP traffic settings but the logs were still showing it blocking wsl outbound connections.

**Comment by raidenii** [comment]
> 
> 
> > Another way of allowing traffic for people use Symantec Endpoint Protection without allowing all IP traffic:
> > Add a firewall rule that allows IP traffic (or traffic you want, like IP protocol 1 (ICMP), 6 (TCP) and 17 (UDP)) on the Hyper-V adapter, both directions. You can find which adapter WSL2 uses in Network & Sharing Center (it will say vEthernet (WSL) in its name).
> 
> How can one do that? Apparently, symantec endpoint protection firewall rules ('security exception risks') should relate to either 1) known risks, 2) files, 3) folders, 4) extensions, or 5) web domains. (these might not be the exact terms: I am roughly translating back from my interface, which is in Portuguese).
> 
> i have tried to make an exception to wsl.exe, under Windows/System32, but it apparently did not work properly. How could one make exceptions only for certain types of IP traffic?
> 
> In my case, internet seems to be working rather properly (I can ping google.com with good results), but ubuntu's servers (for apt get update, for instance) do not work, and http addresses within command line instructions do not work either.

By adding a firewall rule on the WSL network adapter. Not the application settings, but configure firewall rules.

**Comment by ultramaks** [comment]
Can anyone tall me what exact steps should be performed to allow WSL2 traffic from WSL2 Ubuntu  to Internet trough primary host Win10 workstation?
I tried [this](https://superuser.com/questions/1454975/is-it-possible-to-assign-windows-firewall-rules-to-specific-adapters) specifying an IP address of the WSL adapter that I received from ipconfig /all but had no luck
Am i missing something?

**Comment by shahawi-sumup** [comment]
Thanks @anthonysomerset , I could also ping google.com but `apt update` or `git clone` fails :(
So I checked my Avast firewall and it was it :(.
More here: https://businesshelp.avast.com/Content/Products/AfB_Antivirus/ConfiguringSettings/NetworkSettings.htm

Goto Protection -> Firewall -> Settings -> Enable "Internet Connection sharing mode"
Now internet works!

I don't know if this option is completely safe or not, will try to find out more.

**Comment by ExtremeGTX** [comment]
Thanks @anthonysomerset , I could also ping google.com but apt update or git clone fails :(
So I checked my Avast firewall and it was it :(.
More here: businesshelp.avast.com/Content/Products/AfB_Antivirus/ConfiguringSettings/NetworkSettings.htm

Goto Protection -> Firewall -> Settings -> Enable "Internet Connection sharing mode"
Now internet works!

I don't know if this option is completely safe or not, will try to find out more.

**Comment by k0psutin** [comment]
I got the network working in Ubuntu 20.04 WSL2 after I unchecked IPv6 from my WLAN and VPN properties and made few changes to `.conf`-files.

For the `.conf`-files I used the solution found in [https://github.com/microsoft/WSL/issues/5420#issuecomment-646479747](https://github.com/microsoft/WSL/issues/5420#issuecomment-646479747) to stop WSL from rewriting the nameservers.
```
sudo rm /etc/resolv.conf
sudo bash -c 'echo "nameserver 8.8.8.8" > /etc/resolv.conf'
sudo bash -c 'echo "[network]" > /etc/wsl.conf'
sudo bash -c 'echo "generateResolvConf = false" >> /etc/wsl.conf'
sudo chattr +i /etc/resolv.conf
```

These steps worked for me and I hope it helps others as well. Now I have network connection in Ubuntu everytime I launch it.

**Comment by rabbagliettiandrea** [comment]
@sboonyakiatACR hello mate, I'm having issue with SEP & WSL2 but I've the GUI locked by my corporation. Could you please give me your policy.xml that you could generate through SEP -> Troubleshooting -> Export profile so I can import it with the firewall exception on. Thank you so much, you can find my contact detail in my github profile, thanks!

**Comment by uscwifi** [comment]
Hi，我也遇到了这个问题，我是在更新了windows 6月份的补丁之后发现的这个问题，我参考这篇文章：[ipconfig shows only "Windows IP Configuration" after upgrade to Win10](https://answers.microsoft.com/en-us/windows/forum/all/ipconfig-shows-only-windows-ip-configuration-after/9b48e6c2-3a88-497b-98b5-d7ecab27b478?page=2&auth=1)
![image](https://user-images.githubusercontent.com/42139780/123964568-1f8d6e00-d9e6-11eb-9772-b822d2532498.png)
我卸载掉电脑上的checkpoint vpn然后重启电脑后，我的ubuntu子系统就正常了，我真是个幸运儿！:yum:

**Comment by brettcave** [comment]
I had a similar issue - to narrow down the problem, I avoided DNS and worked with IPs - so it was a routing / nat related issue (and because I couldn't reach my windows local network from the virtual network, DNS fails, but it's not the root cause). 

Problem description: My Windows host is on network a.0.0.0/24 with a default gateway of a.0.0.1. from windows. I can ping from Windows to default gateway, or out onto the network (e.g. 8.8.8.8). I had full network functionality in WSL1 (Ubuntu 20). After WSL2 upgrade, networking was lost. From Ubuntu, was not able to ping a.0.0.1, my local network gateway (and therefore any external IP like 8.8.8.8).

Resolution: 
1. I had WSL + Virtual Machine Platform installed in Windows Features, but did not have Hyper-V components. I added all Hyper-V components and restarted.
2. Still not able to reach local gateway. The solution posted by @Karlheinzniebuhr did the trick, but a full Network Reset was not needed (I didn't want to reconfigure VPNs and other network settings that I needed). To confirm the commands, executed from an Admin powershell / command prompt: 

```
wsl --shutdown
netsh winsock reset
netsh int ip reset all
netsh winhttp reset proxy
ipconfig /flushdns
```
Reboot. 

After that, ping to local gateway confirmed, plus ping to external IP was possible, but DNS was not working. 

In Ubuntu, /etc/resolv.conf was autogenerated with a virtual network address (172.31 - the same as the WSL gateway, i.e. virtual network adapter on the Windows host). Am guessing that the DNS proxy or forwarding there was not working - I didn't resolve this by getting this dns forward / masq or whatever it is to work, but used the `/etc/wsl.conf` method described above to disable auto resolv config. My router / local network gateway (a.0.0.1) is a DNS forwarder, so I set up resolv.conf to use that. Not sure yet if this would affect any WSL to WSL resolution that i might need. 

```
$ sudo -i
# cat >/etc/wsl.conf <<-EOF
[network]
generateResolvConf = false
EOF
# mv /etc/resolv.conf /etc/resolv.conf-autogen-symlink
# echo nameserver a.0.0.1 > /etc/resolv.conf
# host www.google.com
www.google.com has address x.y.z.b
```

1 other item worth noting, not sure if it impacted this, was that I was upgrading to WSL2 in order to get docker working in Linux. Not sure if the docker networking affected this in any way, I didn't verify networking after WSL2 before Docker installation.

HTH, happy networking.

**Comment by brettcave** [comment]
p.s. if anyone has a solution to get the generated resolv.conf working, would be great to use this so I don't have to maintain this myself - didn't see a fix or root cause on this.

**Comment by 108806** [comment]
Nothing works for me, I have looked at the other threads too. I am using win 10.0.19043.1081.
I had to go back to the older WSL1 and older Linux distro, openSUSE-15.2 is the only one fully working on my machine currently, everything else has some DNS issues, it can ping 8.8.8.8 but can't ping google.com.

**Comment by flashadvocate** [comment]
Crazy idea, and it actually worked...

Why not just run VPN on the VM itself?

```
apt-get install openconnect
openconnect vpn.domain.com
# <CTRL+Z>
# bg
```

**Comment by lo-w** [comment]
hello, we got the same issue.
the SEP(Symantec Endpoint Protection) blocks the traffic from WSL2. 
but we cannot modify the SEP settings because it's blocked by the organization.
do we having any other WA to put the way forward, thanks.
![image](https://user-images.githubusercontent.com/40700499/126744135-97170c9e-7f47-4e54-9183-09c44c2e7387.png)
![image](https://user-images.githubusercontent.com/40700499/126744297-413135aa-a457-4500-a30a-0f808bbfaef6.png)

**Comment by y1rn** [comment]
> hello, we got the same issue.
> the SEP(Symantec Endpoint Protection) blocks the traffic from WSL2.
> but we cannot modify the SEP settings because it's blocked by the organization.
> do we having any other WA to put the way forward, thanks.
> ![image](https://user-images.githubusercontent.com/40700499/126744135-97170c9e-7f47-4e54-9183-09c44c2e7387.png)
> ![image](https://user-images.githubusercontent.com/40700499/126744297-413135aa-a457-4500-a30a-0f808bbfaef6.png)

same issue

**Comment by lo-w** [comment]
Hi
update the latest status about who using SEP(Symantec Endpoint Protection) having this issue.
we using geek(uninstall tool) to remove the SEP totally, now everything works as expected.
BR//lo

**Comment by monigenomi** [comment]
I think what helped for me was:

- Uninstall all Ubuntu versions
- Disable "Enable integration with my default WSL distro" in Docker's settings (Resources -> WSL Integration)
- Restart computer
- Install Ubuntu 20.04 LTS
- Check "Ubuntu-20.04" in "Enable integration with additional distros" in mentioned Docker's settings panel (**do not** check "Enable integration with my default WSL distro"
- Click "Apply & Restart"
- Optionally restart computer again

**Comment by jramiresbrito** [comment]
> Fixed by applying this [#5336 (comment)](https://github.com/microsoft/WSL/issues/5336#issuecomment-668773304)
> 
> ============================================================================= FIX WSL2 NETWORKING IN WINDOWS 10 cmd as admin: wsl --shutdown netsh winsock reset netsh int ip reset all netsh winhttp reset proxy ipconfig /flushdns
> 
> Windows Search > Network Reset
> 
> Restart Windows

This worked like a charm for me in W11 and WSL2

**Comment by felipecrs** [comment]
For those using Symantec Endpoint Protection, I shared a workaround at https://github.com/MicrosoftDocs/WSL/pull/1046#issuecomment-1010092796.

Perhaps it could be useful for other firewalls as well.

**Comment by mabasic** [comment]
For anyone still struggling I have found that allowing incoming traffic on the public profile in the firewall instantly makes DNS resolution work in WSL without having to make any changes to /etc/resolv.conf or having to create /etc/wsl.conf file.

Windows Defender Firewall with Advanced Security -> Windows Defender Firewall Properties -> Public Profile -> Inbound Connections set to Allow.

My company uses Firewall to block certain ports which causes this issue. When I set this to allow all inbound connections DNS resolution inside WSL starts working instantly.

**Comment by Nazeim** [comment]
> For anyone still struggling I have found that allowing incoming traffic on the public profile in the firewall instantly makes DNS resolution work in WSL without having to make any changes to /etc/resolv.conf or having to create /etc/wsl.conf file.
> 
> Windows Defender Firewall with Advanced Security -> Windows Defender Firewall Properties -> Public Profile -> Inbound Connections set to Allow.
> 
> My company uses Firewall to block certain ports which causes this issue. When I set this to allow all inbound connections DNS resolution inside WSL starts working instantly.

IT WORKED, without even a reboot.

**Comment by knutov** [comment]
> Windows Defender Firewall with Advanced Security -> Windows Defender Firewall Properties -> Public Profile -> Inbound Connections set to Allow.

works for dns and http[s], but not for ssh from wsl2 to external network

**Comment by knutov** [comment]
`sudo ip link set dev eth0 mtu 1350` fixes all other problems (see https://github.com/microsoft/WSL/issues/4253 for details)

**Comment by ManalLiaquat** [comment]
Nothing works for me, so I gave my windows user Administrative access and reinstall the WSL and then It works like a charm

**Comment by LouDnl** [comment]
> For anyone still struggling I have found that allowing incoming traffic on the public profile in the firewall instantly makes DNS resolution work in WSL without having to make any changes to /etc/resolv.conf or having to create /etc/wsl.conf file.
> 
> Windows Defender Firewall with Advanced Security -> Windows Defender Firewall Properties -> Public Profile -> Inbound Connections set to Allow.
> 
> My company uses Firewall to block certain ports which causes this issue. When I set this to allow all inbound connections DNS resolution inside WSL starts working instantly.

Fixed it for me. Anyone create a rule to not have to allow all incoming traffic?

**Comment by warengonzaga** [comment]
For those using ESET Antivirus, all you have to do is unblock the connection below.
<img width="600" alt="image" src="https://github.com/microsoft/WSL/assets/15052701/62724558-5daf-47f4-9843-0cb04e29effb">

Open your ESET settings > Resolved blocked communication
<img width="960" alt="image" src="https://github.com/microsoft/WSL/assets/15052701/e60f7aef-f1ec-414a-a170-24663037d4f8">

And it should work!

**Comment by merovingian** [comment]
> @j0057 Thank you, you have helped me solve my problem that I have been debugging for over 3 months! Those NAT rules that I had there (probably from previous VM's) were conflicting with WSL2, so I removed them with `Remove-NetNat` and boom! WSL2 networking is fixed (in my case)!

Thank you! Remove-NetNat worked for me! This has been bugging me for three days!

**Comment by wirytiox** [comment]
hello, i found a solution. https://github.com/dorssel/usbipd-win/issues/714
if you run it in mirrored mode you should not have the same issue.

**Comment by hilarycheng** [comment]
I am facing this problem in company computer which is Symantec Endpoint Protection. It is no way to enable to configure SEP to enable WSL 2 to access Internet.

Since WSL 1 is working fine with SEP. As Microsoft allow both WSL 1 and WSL 2 running at the same time. I just install a squid proxy at WSL 1. Configurate the WSL2 Ubuntu to use WSL1 as HTTP / HTTPS Proxy. It is working fine. But you have to manually configure for each application.

**Comment by MartinBriand** [comment]
Hi,
I have had the same issue when switching from WSL1 to WSL2 on my company laptop. It was (I think but I am not sure since I need Administrator privilege to check that) due to Symantec Endpoint Protection that is installed on the Windows side of my machine. Most of the proposed solutions require administrator rights on Windows, which I don’t have.

I was pretty frustrated and really needed WSL2 in order to use VScode on windows while having a Linux development environment.

HOWEVER, serendipity has always been a good friend: in the "Windows Subsystem for Linux Settings" app, under "Network", using the "VirtioProxy" "Networking mode" simply solved the issue.

Yeah, really, just that simple.

I have no idea of what is happening under the hood since there is nothing on the WSL official documentation. Whatever, it works for me and that is all I need.

I hope it will work for you and that Microsoft will explain a bit more about this "VirtioProxy Networking mode".

Enjoy your day,
Martin

**Comment by sombek** [comment]
> Our team reports about verified workaround https://github.com/sakai135/wsl-vpnkit with window 10 build `1909`, that you could try

After one year of interval debugging this is the only solution worked for us.
